### ISSUE TITLE
nginx not able to resolve extra_hosts in Dev Containers

### ISSUE THREAD

**Issue by f1yn** [microsoft/vscode-remote-release#7730]
Type: <b>Bug</b>

In GitHub Codespaces, I'm able to connect to services running in the top level container (not host) using the `extra_hosts` and `'host.docker.internal:host-gateway'`

Locally however, even using dev containers - the networking stack seems to break. I can ping `host.docker.internal`, curl `host.docker.internal`, and see the correct IP address in `/etc/hosts` in the container - but nginx does NOT connect to it when using Dev Containers.

```
2022/12/17 07:31:11 [error] 12#12: *36 host.docker.internal could not be resolved (3: Host not found), client: 127.0.0.1, server: , request: "GET /embed HTTP/1.0", host: "[secret host shhhh]"
```

In case it helps, I'm running Vs Code inside of a VM, and the Dev Container is inside the VM.

Extension version: 0.266.1
VS Code version: Code 1.74.1 (1ad8d514439d5077d2b0b7ee64d2ce82a9308e5a, 2022-12-14T10:35:03.979Z)
OS version: Linux x64 6.0.6-76060006-generic
Modes:
Sandboxed: No
Remote OS version: Linux x64 6.0.6-76060006-generic

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i9-10850K CPU @ 3.60GHz (10 x 3599)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: disabled_software<br>video_encode: disabled_software<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: disabled_off|
|Load (avg)|0, 1, 1|
|Memory (System)|15.62GB (10.06GB free)|
|Process Argv|--unity-launch --crash-reporter-id 4e598f63-d287-44cd-b916-fc7d5169b9b8|
|Screen Reader|no|
|VM|0%|
|DESKTOP_SESSION|cinnamon|
|XDG_CURRENT_DESKTOP|X-Cinnamon|
|XDG_SESSION_DESKTOP|cinnamon|
|XDG_SESSION_TYPE|x11|

|Item|Value|
|---|---|
|Remote|Dev Container: Node.js|
|OS|Linux x64 6.0.6-76060006-generic|
|CPUs|Intel(R) Core(TM) i9-10850K CPU @ 3.60GHz (10 x 3599)|
|Memory (System)|15.62GB (10.06GB free)|
|VM|0%|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vsreu685:30147344
python383:30185418
vspor879:30202332
vspor708:30202333
vspor363:30204092
vslsvsres303:30308271
pythonvspyl392:30443607
vserr242cf:30382550
pythontb:30283811
vsjup518:30340749
pythonptprofiler:30281270
vsdfh931cf:30280410
vshan820:30294714
vstes263:30335439
vscoreces:30445986
pythondataviewer:30285071
vscod805:30301674
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
cmake_vspar411:30581797
vsaa593cf:30376535
pythonvs932:30410667
cppdebug:30492333
vsclangdc:30486549
c4g48928:30535728
dsvsc012cf:30540253
azure-dev_surveyone:30548225
pyindex848:30577860
nodejswelcome1:30587005
3biah626:30602489
gswce1:30612156
iaj6b796:30613358
f6dab269:30613381
fim-prod:30623723

```

</details>

**Comment by chrmarti** [comment]
Are you running `ping`, `curl` and `nginx` in the same container? Why would they show different results for name resolution?

**Comment by f1yn** [comment]
@chrmarti yep. I've made sure the commands are run in the same instance as the ngnix container using `docker exec -it [id] bash`

I've read in a few places that Docker does multicast DNS trickery with custom hostnames (i.e routing to other docker instnaces within the shared network namespace), and that `ping` and `curl` use local hosts, while ngnix relies on upstream DNS only. This shouldn't matter, as Docker networking should be consistent between Codespaces and Dev Containers, at least in theory.

It's also not the ngnix image, as it's shared across both the x86 codespace instance and local via dev containers.

If you need any additional context, examples or demos please reach out.

**Comment by chrmarti** [comment]
@edgonmsft Is Codespaces doing something to add `/etc/hosts` entries to the DNS used by the dev container?

**Comment by f1yn** [comment]
This is more of a weirdness of Docker Engine than Dev Containers, so I'll close this for now. I appreciate your time @chrmarti, and I'm sorry if I accidentally took you away from other work.

I spent a lot of time trying to figure out why `add-host` DNS queries were failing, but normal ones were not. Inside of the Docker containers themselves (inside of the local Docker instance, as well as Dev Containers) the upstream DNS server claimed to be correctly configured to the Docker DNS resolver `127.0.0.11`.

Previously that same day, I made some modifications to `dnsmasq` on the Docker host when I added some new addresses and aliases that weren't configured for any of my containers yet. I then found out, **Docker (on all levels) was pushing direct DNS queries to the Docker host, which is why I was getting DNS entries for everything except the `extra_hosts`. Nothing in `resolv.conf` in any container would have suggested this, as they were all configured with `127.0.0.11`.**

Unfortunately, no amount of re-configuring of the docker daemon seemed to help - so ultimately I configured additional dnsmasq rules that pointed `host.docker.internal` to the static IP of my workstation - although I could have just used the Docker `host-gateway` IP as well.

Anyways, maybe this will help someone someday - I'm not sure if this is a Docker bug or not, but it definitely threw me in a loop.
